tests/unit/s2n_server_hello_retry_test.c

```diff
@@ -15,15 +15,325 @@
+    /* Retry requests without a supported version extension are not accepted */
+    {
+        struct s2n_config *conf;
+        struct s2n_connection *conn;
+
+        EXPECT_NOT_NULL(conf = s2n_config_new());
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_SUCCESS(s2n_connection_set_config(conn, conf));
+
+        struct s2n_stuffer *io = &conn->handshake.io;
+        conn->server_protocol_version = S2N_TLS13;
+
+        /* protocol version */
+        EXPECT_SUCCESS(s2n_stuffer_write_uint8(io, S2N_TLS12 / 10));
+        EXPECT_SUCCESS(s2n_stuffer_write_uint8(io, S2N_TLS12 % 10));
+
+        /* random data */
+        EXPECT_SUCCESS(s2n_stuffer_write_bytes(io, hello_retry_request_random_test_buffer, S2N_TLS_RANDOM_DATA_LEN));
+
+        /* session id */
+        uint8_t session_id[S2N_TLS_SESSION_ID_MAX_LEN] = {0};
+        EXPECT_SUCCESS(s2n_stuffer_write_uint8(io, S2N_TLS_SESSION_ID_MAX_LEN));
+        EXPECT_SUCCESS(s2n_stuffer_write_bytes(io, session_id, S2N_TLS_SESSION_ID_MAX_LEN));
+
+        /* cipher suites */
+        EXPECT_SUCCESS(s2n_stuffer_write_uint16(io, (0x13 << 8) + 0x01));
+
+        /* no compression */
+        EXPECT_SUCCESS(s2n_stuffer_write_uint8(io, 0));
+
+        EXPECT_FAILURE_WITH_ERRNO(s2n_server_hello_recv(conn), S2N_ERR_BAD_MESSAGE);
+
+        EXPECT_EQUAL(conn->handshake.hello_retry_request, 0);
+
+        EXPECT_SUCCESS(s2n_config_free(conf));
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* Verify the client key share extension properly handles HelloRetryRequests */
+    {
+        struct s2n_connection *server_send_conn;
+        struct s2n_connection *client_recv_conn;
+
+        EXPECT_NOT_NULL(server_send_conn = s2n_connection_new(S2N_SERVER));
+        EXPECT_NOT_NULL(client_recv_conn = s2n_connection_new(S2N_CLIENT));
+
+        struct s2n_stuffer* extension_stuffer = &server_send_conn->handshake.io;
+
+        server_send_conn->secure.server_ecc_evp_params.negotiated_curve = s2n_all_supported_curves_list[0];
+        server_send_conn->secure.client_ecc_evp_params[0].negotiated_curve = s2n_all_supported_curves_list[0];
+        EXPECT_SUCCESS(s2n_server_should_retry(server_send_conn));
+        EXPECT_SUCCESS(s2n_ecc_evp_generate_ephemeral_key(&server_send_conn->secure.client_ecc_evp_params[0]));
+        EXPECT_SUCCESS(s2n_extensions_server_key_share_send(server_send_conn, extension_stuffer));
+
+        S2N_STUFFER_READ_EXPECT_EQUAL(extension_stuffer, TLS_EXTENSION_KEY_SHARE, uint16);
+        /* 4 = S2N_SIZE_OF_EXTENSION_TYPE + S2N_SIZE_OF_EXTENSION_DATA_SIZE */
+        S2N_STUFFER_READ_EXPECT_EQUAL(extension_stuffer, s2n_extensions_server_key_share_send_size(server_send_conn) - 4, uint16);
+
+        client_recv_conn->secure.client_ecc_evp_params[0].negotiated_curve = s2n_all_supported_curves_list[0];
+        EXPECT_SUCCESS(s2n_ecc_evp_generate_ephemeral_key(&client_recv_conn->secure.client_ecc_evp_params[0]));
+
+        /* Parse key share */
```

Comment:
Should this comment be on line 242 instead?

Reply:
My line numbers don't seem to match this at all. I moved the comment right before the call to the extension recv.